mixins/interactive-mixin.js

```diff
@@ -0,0 +1,115 @@
+import { css, html } from 'lit';
+import { findComposedAncestor, isComposedAncestor } from '../helpers/dom.js';
+import { forceFocusVisible, getNextFocusable } from '../helpers/focus.js';
+import { ifDefined } from 'lit/directives/if-defined.js';
+import { LocalizeCoreElement } from '../helpers/localize-core-element.js';
+
+const keyCodes = {
+	ENTER: 13,
+	ESCAPE: 27
+};
+
+export const InteractiveMixin = superclass => class extends LocalizeCoreElement(superclass) {
+
+	static get properties() {
+		return {
+			_hasInteractiveAncestor: { state: true },
+			_interactive: { state: true }
+		};
+	}
+
+	static get styles() {
+		return css`
+			.interactive-container.focus-visible,
+			.interactive-container:focus-visible {
+				border-radius: 6px;
+				outline: 2px solid var(--d2l-color-celestine);
+				outline-offset: 2px;
+			}
+		`;
+	}
+
+	constructor() {
+		super();
+		this._hasInteractiveAncestor = false;
+		this._interactive = false;
+	}
+
+	connectedCallback() {
+		super.connectedCallback();
+
+		const parentGrid = findComposedAncestor(this.parentNode, node => {
+			return ((node.tagName === 'D2L-LIST' && node.grid) || (node.nodeType === Node.ELEMENT_NODE && node.getAttribute('role') === 'grid'));
+		});
+		this._hasInteractiveAncestor = (parentGrid !== null);
+	}
+
+	focus() {
+		if (!this.shadowRoot) return;
+		if (this._hasInteractiveAncestor && !this._interactive) forceFocusVisible(this.shadowRoot.querySelector('.interactive-container'));
+		else this._focusInteractiveContent();
+	}
+
+	_handleInteractiveContentFocusIn() {
+		this._interactive = true;
+	}
+
+	_handleInteractiveContentFocusOut(e) {
+		if (isComposedAncestor(this.shadowRoot.querySelector('.interactive-container-content'), e.relatedTarget)) return;
+		// focus moved out of the interactive content
+		this._interactive = false;
+	}
+
+	async _handleInteractiveEndFocus() {
+		// focus moved to trap-end either forwards from contents or backwards from outside - focus interactive toggle
+		this._interactive = false;
+		await this.updateComplete;
+		this.shadowRoot.querySelector('.interactive-container').focus();
+	}
+
+	async _handleInteractiveKeyDown(e) {
+		if (!this._hasInteractiveAncestor) return;
```

Comment:
I don't think the event handler would have even been wired up if `_hasInteractiveAncestor` were `false`?

Reply:
That's a good point.  I think I might have written this before the conditional rendering was in place.  I think this can be removed.